pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1308,36 +1308,44 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta Predicate
-// of the "pod" on the given "nodeInfo".Node(). It returns three values: 1) whether
-// targetPod matches all the terms and their topologies, 2) whether targetPod
-// matches all the terms label selector and namespaces (AKA term properties),
-// 3) any error.
-func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo, terms []v1.PodAffinityTerm) (bool, bool, error) {
-	if len(terms) == 0 {
-		return false, false, fmt.Errorf("terms array is empty")
-	}
-	props, err := getAffinityTermProperties(pod, terms)
+// podMatchesPodAffinityTerms checks if the "targetPod" matches the "checkAffinity (true:affinity,false:anti-affinity)"
+// of the "pod" on the given "nodeInfo".Node(). It returns four values: 1) whether
+// targetPod matches affnityTerms and their topologies based on the request of affinity or anti-affinity 2) whether targetPod
+// matches all the terms label selector and namespaces (AKA term properties), 3) any error.
+func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo, checkAffinity bool) (bool, bool, error) {
+	matchedTerms, affinityTermsNum, err := getMatchingAffinityTerms(pod, targetPod, checkAffinity)
 	if err != nil {
 		return false, false, err
 	}
-	if !podMatchesAllAffinityTermProperties(targetPod, props) {
-		return false, false, nil
+	termsSelectorMatched := len(matchedTerms) == affinityTermsNum
+	if checkAffinity && !termsSelectorMatched {
+		return false, false, err
 	}
-	// Namespace and selector of the terms have matched. Now we check topology of the terms.
 	targetPodNode, err := c.info.GetNodeInfo(targetPod.Spec.NodeName)
 	if err != nil {
-		return false, false, err
+		return false, termsSelectorMatched, err
 	}
-	for _, term := range terms {
+	if len(matchedTerms) == 0 {
+		return false, false, nil
+	}
+	matchedTermCount := 0
+	// Namespace and selector of the terms have matched. Now we check topology of the terms.
+	for _, term := range matchedTerms {
 		if len(term.TopologyKey) == 0 {
-			return false, false, fmt.Errorf("empty topologyKey is not allowed except for PreferredDuringScheduling pod anti-affinity")
-		}
-		if !priorityutil.NodesHaveSameTopologyKey(nodeInfo.Node(), targetPodNode, term.TopologyKey) {
-			return false, true, nil
+			return false, termsSelectorMatched, fmt.Errorf("empty topologyKey is not allowed except for PreferredDuringScheduling pod anti-affinity")
+		}
+		topologyMatched := priorityutil.NodesHaveSameTopologyKey(nodeInfo.Node(), targetPodNode, term.TopologyKey)
+		if topologyMatched {
+			matchedTermCount++
+			if checkAffinity == false {
+				return true, termsSelectorMatched, nil
+				//For podAffinity, we must check the total number of the affinityTerms, it must be the same as matchedTermCount.
+			} else if checkAffinity && matchedTermCount == affinityTermsNum && termsSelectorMatched {
+				return true, true, nil
+			}
```

Comment:
IMO this can be moved right above L1348 (the final return) to be a standalone `if` block:

```go
// For podAffinity, we must check the total number of the affinityTerms, it must be the same as matchedTermCount.
if checkAffinity && matchedTermCount == affinityTermsNum && termsSelectorMatched {
	return true, true, nil
}
```

And with this change, `matchedTermCount++` can be moved to the end of the `if topologyMatched` block.

Reply:
I did not get you point . If we use other standalong if block . Won't we have two loops for this calculate ? One for anti-affinity calculating all mathedTerms and another one for affinity?